R/knitr-hooks.R

```diff
@@ -49,6 +65,43 @@ install_knitr_hooks <- function() {
     grepl("-setup$", label) || (length(exercise_chunks_for_setup_chunk(label)) > 0)
   }
 
+  # helper function to grab the raw knitr chunk associated with a chunk label
+  get_knitr_chunk <- function(label) {
+    code_query <- paste0("knitr::knit_code$get('", label, "')")
+    # Note: we can get the raw, unevaluated chunk options, for e.g. `exercise=as.logical(1)`
+    eval(parse(text = code_query))
+  }
+
+  # helper function to find all the setup chunks associated with an exercise chunk
+  # it goes up the chain of setup dependencies and returns a list of raw knitr chunks (if any)
+  find_parent_setup_chunks <- function(options, visited = NULL) {
+    # base case: empty setup name means no more setup references
+    if (is.null(options))
+      return(NULL)
+    has_visited <- options$label %in% visited
+    # update visited set
+    visited <- append(visited, options$label)
+    # error out if there is a cycle
+    if (has_visited) {
+      stop("Chained setup chunks form a cycle!\nCycle: ", paste0(visited, collapse = " => "), call. = FALSE)
+    }
+    # check if the chunk with label has another setup chunk associated with it
+    label <- options$exercise.setup
+    code_chunk <- get_knitr_chunk(label)
+    # if the label is mispelled, throw an error to user instead of silently ignoring
+    # which will cause other issues when data dependencies can't be found
+    if (!is.null(label) && is.null(code_chunk))
+      stop(paste0("exercise.setup label '", label, "' not found for exercise '", options$label, "'"))
+    # recurse if the chunk options exist, else return
+    options <- attr(code_chunk, "chunk_opts")
+    setup_chunks <- list()
+    # this check will make sure we exclude the last NULL element (base case)
+    if (!is.null(options))
+      setup_chunks <- list(code_chunk)
+    setup_chunks <- append(setup_chunks, find_parent_setup_chunks(options, visited))
```

Comment:
Overwriting names is usually a bad practice. I also got confused when reading it, thinking it was the original `options` that was sent in
```suggestion
    # check if the chunk with label has another setup chunk associated with it
    setup_label <- options$exercise.setup
    setup_chunk <- get_knitr_chunk(label)
    # if the label is misspelled, throw an error to user instead of silently ignoring
    # which will cause other issues when data dependencies can't be found
    if (!is.null(setup_label) && is.null(setup_chunk))
      stop(paste0("exercise.setup label '", setup_label, "' not found for exercise '", options$label, "'"))
    setup_options <- attr(code_chunk, "chunk_opts")

    # recurse
    parent_setup_chunks <- find_parent_setup_chunks(options, visited)
    setup_chunks <- c(parent_setup_chunks, setup_chunk)
```

By using `c()` if `parent_setup_chunks` is `NULL`, then it'll just be `setup_chunk`.